src/backend/columnar/columnar_customscan.c

```diff
@@ -146,6 +160,18 @@ columnar_customscan_init()
 		PGC_USERSET,
 		GUC_NO_SHOW_ALL,
 		NULL, NULL, NULL);
+	DefineCustomIntVariable(
+		"columnar.max_custom_scan_paths",
+		gettext_noop("Maximum number of custom scan paths to generate "
+					 "for a columnar table when planning."),
+		NULL,
+		&ColumnarMaxCustomScanPaths,
+		64,
```

Comment:
Can we have a few tests where query plan with different settings of `columnar.max_custom_scan_path` changes thanks to having a greater search space ?

Reply:
e.g.:

```sql
set columnar.max_custom_scan_paths to 1000;
EXPLAIN (analyze on, costs off, timing off, summary off)
  SELECT * FROM coltest c1, coltest c2, coltest c3, coltest c4 WHERE
    c1.id = c2.id and c1.id = c3.id and c1.id = c4.id and c1.x1 = c2.id and c2.x1 = c1.x1 and c2.x1 = c4.x3;
┌────────────────────────────────────────────────────────────────────────────────────┐
│                                     QUERY PLAN                                     │
├────────────────────────────────────────────────────────────────────────────────────┤
│ Nested Loop (actual rows=1 loops=1)                                                │
│   ->  Nested Loop (actual rows=1 loops=1)                                          │
│         ->  Nested Loop (actual rows=1 loops=1)                                    │
│               ->  Custom Scan (ColumnarScan) on coltest c1 (actual rows=1 loops=1) │
│                     Filter: (id = x1)                                              │
│                     Rows Removed by Filter: 19999                                  │
│               ->  Custom Scan (ColumnarScan) on coltest c2 (actual rows=1 loops=1) │
│                     Filter: ((id = x1) AND (c1.id = id))                           │
│                     Rows Removed by Filter: 999                                    │
│                     Columnar Chunk Group Filters: (c1.id = id)                     │
│                     Columnar Chunk Groups Removed by Filter: 19                    │
│         ->  Custom Scan (ColumnarScan) on coltest c3 (actual rows=1 loops=1)       │
│               Filter: (c1.id = id)                                                 │
│               Rows Removed by Filter: 999                                          │
│               Columnar Chunk Group Filters: (c1.id = id)                           │
│               Columnar Chunk Groups Removed by Filter: 19                          │
│   ->  Custom Scan (ColumnarScan) on coltest c4 (actual rows=1 loops=1)             │
│         Filter: ((id = x3) AND (c1.id = id))                                       │
│         Rows Removed by Filter: 999                                                │
│         Columnar Chunk Group Filters: (c1.id = id)                                 │
│         Columnar Chunk Groups Removed by Filter: 19                                │
└────────────────────────────────────────────────────────────────────────────────────┘
(21 rows)

Time: 22.798 ms
```

```sql
set columnar.max_custom_scan_paths to 2;
EXPLAIN (analyze on, costs off, timing off, summary off)
  SELECT * FROM coltest c1, coltest c2, coltest c3, coltest c4 WHERE
    c1.id = c2.id and c1.id = c3.id and c1.id = c4.id and c1.x1 = c2.id and c2.x1 = c1.x1 and c2.x1 = c4.x3;
┌────────────────────────────────────────────────────────────────────────────────────┐
│                                     QUERY PLAN                                     │
├────────────────────────────────────────────────────────────────────────────────────┤
│ Nested Loop (actual rows=1 loops=1)                                                │
│   Join Filter: (c1.id = c4.id)                                                     │
│   ->  Nested Loop (actual rows=1 loops=1)                                          │
│         Join Filter: (c1.id = c3.id)                                               │
│         Rows Removed by Join Filter: 19999                                         │
│         ->  Nested Loop (actual rows=1 loops=1)                                    │
│               Join Filter: (c1.id = c2.id)                                         │
│               ->  Custom Scan (ColumnarScan) on coltest c1 (actual rows=1 loops=1) │
│                     Filter: (id = x1)                                              │
│                     Rows Removed by Filter: 19999                                  │
│               ->  Custom Scan (ColumnarScan) on coltest c2 (actual rows=1 loops=1) │
│                     Filter: (id = x1)                                              │
│                     Rows Removed by Filter: 19999                                  │
│         ->  Custom Scan (ColumnarScan) on coltest c3 (actual rows=20000 loops=1)   │
│   ->  Custom Scan (ColumnarScan) on coltest c4 (actual rows=1 loops=1)             │
│         Filter: (id = x3)                                                          │
│         Rows Removed by Filter: 19999                                              │
└────────────────────────────────────────────────────────────────────────────────────┘
(17 rows)

Time: 31.822 ms
```

Reply:
Done. Thank you for the test case -- I modified it slightly to avoid excessive debug info.